bin/trait_mapping/traits_to_zooma.py

```diff
@@ -0,0 +1,170 @@
+            else:
+                output_file.write("\tZOOMA_MAPPING_FAILED")
+            output_file.write("\n")
+
+
+
+def read_traits(filepath):
+    traits = []
+    with open(filepath) as f:
+        for line in f:
+            line = line.rstrip()
+            line_list = line.split("\t")
+            name = line_list[0]
+            frequency = line_list[1]
+            traits.append(Trait(name, frequency))
+    return traits
+
+
+def get_ontology_mappings(trait, filters):
+    '''
+    First get the URI, label from a selected source, confidence and source:
+    http://snarf.ebi.ac.uk:8580/spot/zooma/v2/api/services/annotate?propertyValue=intellectual+disability
+    Then the ontology label to replace the label from a source:
+    http://www.ebi.ac.uk/ols/api/terms?iri=http%3A%2F%2Fwww.ebi.ac.uk%2Fefo%2FEFO_0003847
+    '''
+    url = build_zooma_query(trait.name, filters)
+    retry_count = 4
+    for retry_num in range(retry_count):  # retries
+        try:
+            json_response_1 = requests.get(url).json()
+            break
+        except json.decoder.JSONDecodeError as e:
+            print("attempt {}: decode error for request with url {}".format(retry_num, url))
+            if retry_num == retry_count - 1:
+                print("error on last attempt, skipping")
+                return None
+
+    mappings = get_mappings_for_trait(json_response_1, trait)
+
+    for mapping in mappings:
+        try:
+            label = get_ontology_label_from_ols(mapping.uri)
+            # If no label is returned (shouldn't really happen) keep the existing one
+            if label:
+                mapping.label = label
+                mapping.ols_label = 1
+        except:
+            print(
+                "Couldn't retrieve ontology label from OLS for trait '{}', will use the one from Zooma".format(
+                    trait.name))
+
+    return mappings
+
+
+def build_zooma_query(trait_name, filters):
+    url = "https://www.ebi.ac.uk/spot/zooma/v2/api/services/annotate?propertyValue={}".format(trait_name)
+    url_filters = [
+                    "required:[{}]".format(filters["required"]),
+                    "ontologies:[{}]".format(filters["ontologies"]),
+                    "preferred:[{}]".format(filters["preferred"])
+                  ]
+    url += "&filter={}".format(",".join(url_filters))
+    return url
+
+
+def get_mappings_for_trait(zooma_response, trait):
+    mappings = []
+    for result in zooma_response:
+        uri = ",".join(result["semanticTags"])
+        label = result["annotatedProperty"]["propertyValue"]
+        confidence = result["confidence"]
+        source_name = result["derivedFrom"]["provenance"]["source"]["name"]
+        mappings.append(OntologyMapping(uri, label, confidence, source_name))
+    return mappings
+
+
+def get_ontology_label_from_ols(uri_mapping):
+    url = build_ols_query(uri_mapping)
+
+    retry_count = 4
+    for retry_num in range(retry_count):  # retries
```

Comment:
This will retry 4 times if no defining ontology is found. Since most of the retry logic is the same, I would suggest creating a function for that, which receives another function as argument to run some logic like the `for term in json_response` loop, and if there is a return value then it just breaks the loop.